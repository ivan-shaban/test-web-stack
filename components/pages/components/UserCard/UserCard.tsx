import {FC} from 'react'
import styles from './UserCard.module.scss'
import {ClientUser} from '../../../../types/client'

export interface Props {
    readonly user: ClientUser
}

export const UserCard: FC<Props> = ({user}) => {
    const createAt = new Date(user.dob).toLocaleDateString()
    return (
        <div className={styles.base}>
            <img
                className={styles.editIcon}
                src="/static/icons/edit-icon.svg"
            />
            <img
                className={styles.avatar}
                src={user.image}
                alt={`${user.name}'s avatar`}
            />
            <div className={styles.credentials}>
                <p className={styles.title}>{user.name}</p>
                <p className={styles.createdTitle}>created <span className={styles.createdValue}>{createAt}</span></p>
            </div>
            <p className={styles.description}>{user.description}</p>
        </div>
    )
}
