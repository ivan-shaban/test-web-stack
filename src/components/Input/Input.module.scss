.base {
    height: 64px;
    padding: 15px;
    font-size: 24px;
    border-radius: 8px;
    border-color: rgba(0, 0, 0, 0.1);

    &:focus {
        outline: none !important;
    }

    &::placeholder {
        font-family: 'Source Sans Pro', sans-serif;

        color: #000;
    }
}
