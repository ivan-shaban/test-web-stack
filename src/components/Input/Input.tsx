import {
    DetailedHTMLProps,
    FC,
    InputHTMLAttributes,
    memo,
} from 'react'
import classNames from 'classnames'
import styles from './Input.module.scss'

export interface Props extends DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>{
}

export const Input: FC<Props> = memo(({className, children, ...restProps}) => {
    const baseClasses = classNames(styles.base, className)
    return (
        <input className={baseClasses} {...restProps}>{children}</input>
    )
})
