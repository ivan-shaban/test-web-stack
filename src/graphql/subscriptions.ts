/* tslint:disable */
/* eslint-disable */
// this is an auto generated file. This will be overwritten

export const onCreateUser = /* GraphQL */ `
  subscription OnCreateUser {
    onCreateUser {
      id
      name
      dob
      image
      address
      description
      createdAt
      updatedAt
    }
  }
`;
export const onUpdateUser = /* GraphQL */ `
  subscription OnUpdateUser {
    onUpdateUser {
      id
      name
      dob
      image
      address
      description
      createdAt
      updatedAt
    }
  }
`;
export const onDeleteUser = /* GraphQL */ `
  subscription OnDeleteUser {
    onDeleteUser {
      id
      name
      dob
      image
      address
      description
      createdAt
      updatedAt
    }
  }
`;
