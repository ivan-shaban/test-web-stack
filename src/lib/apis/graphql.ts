export const getAllUsers = /* GraphQL */ `
    query MyQuery {
        listUsers {
            items {
                id
                address
                dob
                name
            }
        }
    }
`
