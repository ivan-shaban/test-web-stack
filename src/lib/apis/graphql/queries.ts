/* tslint:disable */
/* eslint-disable */
// this is an auto generated file. This will be overwritten

export const getUser = /* GraphQL */ `
  query GetUser($id: ID!) {
    getUser(id: $id) {
      id
      name
      dob
      image
      address
      description
      createdAt
      updatedAt
      owner
    }
  }
`;
export const listUsers = /* GraphQL */ `
  query ListUsers(
    $filter: ModelUserFilterInput
    $limit: Int
    $nextToken: String
  ) {
    listUsers(filter: $filter, limit: $limit, nextToken: $nextToken) {
      items {
        id
        name
        dob
        image
        address
        description
        createdAt
        updatedAt
        owner
      }
      nextToken
    }
  }
`;
