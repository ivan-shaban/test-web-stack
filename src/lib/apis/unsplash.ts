import { createApi } from 'unsplash-js';
import nodeFetch from 'node-fetch';

export const unsplashAPI = createApi({
  accessKey: process.env.UNSPLASH_ACCESS_KEY!,
  // @ts-ignore
  fetch: nodeFetch,
});
