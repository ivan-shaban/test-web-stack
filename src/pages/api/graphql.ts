import 'reflect-metadata'
import {ApolloServer} from 'apollo-server-micro'
import Cors from 'micro-cors'
import {PrismaClient} from '@prisma/client'
import prisma from '../../lib/prisma'
import {buildSchema} from 'type-graphql'
import {crudResolvers} from 'graphql/generated/type-graphql'
import path from 'path'

export type Context = {
    prisma: PrismaClient
    userId?: number
}

const cors = Cors()
let apolloServer: ApolloServer

const getServer = async () => {
    if (!apolloServer) {
        const context = (): Context => ({prisma})
        const schema = await buildSchema({
            resolvers: crudResolvers,
            emitSchemaFile: path.resolve(__dirname, '..', '..', '..', '..', 'generated-schema.graphql'),
            validate: false,
        })
        apolloServer = new ApolloServer({
            schema,
            context,
        })

        await apolloServer.start()
    }

    return apolloServer
}

export default cors(async function handler(req, res) {
    if (req.method === 'OPTIONS') {
        res.end()
        return false
    }
    const server = await getServer()
    await server.createHandler({path: '/api/graphql'})(req, res)
})

export const config = {
    api: {
        bodyParser: false,
    },
}
