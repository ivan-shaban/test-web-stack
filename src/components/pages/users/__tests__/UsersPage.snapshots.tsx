import React, {useState as useStateOriginal} from 'react'
import {
    Props,
    UsersPageOriginal,
} from '../UsersPage'
import renderer from 'react-test-renderer'
import {User} from 'graphql/generated/type-graphql/models/User'
import {GET_ALL_USERS_QUERY} from '../../../../lib/apis/graphql'
import {mocked} from 'jest-mock'
import {UserCard} from '../components/UserCard/UserCard'
import {Input} from '../../../Input/Input'

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useState: jest.fn(),
}))
jest.mock('../components/EditUserInfoOverlay/EditUserInfoOverlay', () => ({EditUserInfoOverlay: 'EditUserInfoOverlay'}))
jest.mock('../components/UserCard/UserCard', () => ({UserCard: 'UserCard'}))
jest.mock('../../../Input/Input', () => ({Input: 'Input'}))
jest.mock('../../../Button/Button', () => ({Button: 'Button'}))
jest.mock('../../../Layout/Layout', () => ({Layout: 'Layout'}))

const useState = mocked(useStateOriginal as <T>(arg: T) => [T, Function])
const setFilterValue = jest.fn()
const setUserToEdit = jest.fn()
const setHasScroll = jest.fn()
const setLoadingState = jest.fn()
const useQuery = jest.spyOn(require('@apollo/client'), 'useQuery')
const useRouter = jest.spyOn(require('next/router'), 'useRouter')
const defaultProps: Props = {}
const users: User[] = [
    {
        id: 91,
        name: 'Leanne Grahama',
        // @ts-ignore
        dob: '1990-01-01T00:00:00.000Z',
        address: 'minsk',
        image: 'https://images.unsplash.com/photo-1626038135427-bd4eb8193ba5?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop',
        description: 'Multi-layered client-server neural-net',
    },
    {
        id: 92,
        name: 'Mrs. Dennis Schulist',
        // @ts-ignore
        dob: '1990-01-01T00:00:00.000Z',
        address: 'South Christy, Norberto Crossing',
        image: 'https://images.unsplash.com/photo-1601455763557-db1bea8a9a5a?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop',
        description: 'Synchronised bottom-line interface',
    },
]

const createTree = (props?: Partial<Props>) => {
    return renderer.create(<UsersPageOriginal {...defaultProps} {...props}/>).toJSON()
}
const originalError = console.error

beforeAll(() => {
    const ignoredWarnings = [
        /Warning.*is using incorrect casing. Use PascalCase for React components, or lowercase for HTML elements./,
        /Warning: The tag .* is unrecognized in this browser. If you meant to render a React component, start its name with an uppercase letter./,
        /Warning: React does not recognize the .* prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase .* instead. If you accidentally passed it from a parent component, remove it from the DOM element./,
    ]
    console.error = (...args) => {
        if (ignoredWarnings.some((warningTemplate) => warningTemplate.test(args[0]))) {
            return
        }
        originalError.call(console, ...args)
    }
})

beforeEach(() => {
    useQuery.mockImplementation(() => ({
        data: null,
        fetchMore: jest.fn().mockResolvedValue({}),
    }))
    useRouter.mockImplementation(() => ({
        query: {page: 1},
        push: jest.fn().mockResolvedValue({}),
    }))
})

afterEach(() => {
    useState.mockClear()
    setFilterValue.mockClear()
    setUserToEdit.mockClear()
    setHasScroll.mockClear()
    setLoadingState.mockClear()
    useRouter.mockClear()
    useQuery.mockClear()
})

afterAll(() => {
    console.error = originalError
})

it('default', () => {
    useState.mockImplementationOnce((init) => [init, setFilterValue])
    useState.mockImplementationOnce((init) => [init, setUserToEdit])
    useState.mockImplementationOnce((init) => [init, setHasScroll])
    useState.mockImplementationOnce((init) => [init, setLoadingState])

    const tree = createTree()
    expect(tree).toMatchSnapshot()
})

it('should render cards', function () {
    useState.mockImplementationOnce((init) => [init, setFilterValue])
    useState.mockImplementationOnce((init) => [init, setUserToEdit])
    useState.mockImplementationOnce((init) => [init, setHasScroll])
    useState.mockImplementationOnce((init) => [init, setLoadingState])
    useQuery.mockImplementationOnce(() => ({
        data: {
            users,
        },
    }))

    const tree = createTree()
    expect(tree).toMatchSnapshot()
    expect(useQuery).toHaveBeenCalledWith(GET_ALL_USERS_QUERY, {
        errorPolicy: 'all',
        notifyOnNetworkStatusChange: true,
        variables: {
            // @ts-ignore
            take: 1 * process.env.NEXT_PUBLIC_USERS_PER_PAGE,
        },
    })
})

it('should render overlay', function () {
    useState.mockImplementationOnce((init) => [init, setFilterValue])
    useState.mockImplementationOnce((_init) => [users[0], setUserToEdit])
    useState.mockImplementationOnce((init) => [init, setHasScroll])
    useState.mockImplementationOnce((init) => [init, setLoadingState])
    useQuery.mockImplementationOnce(() => ({
        data: {
            users,
        },
    }))

    const tree = createTree()
    expect(tree).toMatchSnapshot()
})

it('should render overlay with scroll', function () {
    useState.mockImplementationOnce((init) => [init, setFilterValue])
    useState.mockImplementationOnce((_init) => [users[0], setUserToEdit])
    useState.mockImplementationOnce((_init) => [true, setHasScroll])
    useState.mockImplementationOnce((init) => [init, setLoadingState])
    useQuery.mockImplementationOnce(() => ({
        data: {
            users,
        },
    }))

    const tree = createTree()
    expect(tree).toMatchSnapshot()
})

it('should render "loading" state', () => {
    useState.mockImplementationOnce((init) => [init, setFilterValue])
    useState.mockImplementationOnce((init) => [init, setUserToEdit])
    useState.mockImplementationOnce((init) => [init, setHasScroll])
    useState.mockImplementationOnce((_init) => [true, setLoadingState])
    useQuery.mockImplementationOnce(() => ({
        data: {
            users,
        },
    }))

    const tree = createTree({isDisabled: true})
    expect(tree).toMatchSnapshot()
})
