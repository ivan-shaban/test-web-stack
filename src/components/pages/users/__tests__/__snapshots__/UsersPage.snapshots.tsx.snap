// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
<Layout
  className="layout"
  isAbsolute={true}
>
  <div
    className="base"
  >
    <header
      className="header"
    >
      <p
        className="title"
      >
        Users list
      </p>
      <Input
        className="searchInput"
        onChange={[Function]}
        placeholder="Search..."
        type="search"
        value=""
      />
    </header>
    <main
      className="cardList"
    />
    <footer
      className="footer"
    >
      <Button
        id="loadMoreButton"
        isDisabled={true}
        onClick={[Function]}
      >
        Loading...
      </Button>
    </footer>
  </div>
</Layout>
`;

exports[`should render "loading" state 1`] = `
<Layout
  className="layout"
  isAbsolute={true}
>
  <div
    className="base"
  >
    <header
      className="header"
    >
      <p
        className="title"
      >
        Users list
      </p>
      <Input
        className="searchInput"
        onChange={[Function]}
        placeholder="Search..."
        type="search"
        value=""
      />
    </header>
    <main
      className="cardList"
    >
      <UserCard
        isDisabled={true}
        onClick={[Function]}
        user={
          Object {
            "address": "minsk",
            "description": "Multi-layered client-server neural-net",
            "dob": "1990-01-01T00:00:00.000Z",
            "id": 91,
            "image": "https://images.unsplash.com/photo-1626038135427-bd4eb8193ba5?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
            "name": "Leanne Grahama",
          }
        }
      />
      <UserCard
        isDisabled={true}
        onClick={[Function]}
        user={
          Object {
            "address": "South Christy, Norberto Crossing",
            "description": "Synchronised bottom-line interface",
            "dob": "1990-01-01T00:00:00.000Z",
            "id": 92,
            "image": "https://images.unsplash.com/photo-1601455763557-db1bea8a9a5a?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
            "name": "Mrs. Dennis Schulist",
          }
        }
      />
    </main>
    <footer
      className="footer"
    >
      <Button
        id="loadMoreButton"
        isDisabled={true}
        onClick={[Function]}
      >
        Loading...
      </Button>
    </footer>
  </div>
</Layout>
`;

exports[`should render cards 1`] = `
<Layout
  className="layout"
  isAbsolute={true}
>
  <div
    className="base"
  >
    <header
      className="header"
    >
      <p
        className="title"
      >
        Users list
      </p>
      <Input
        className="searchInput"
        onChange={[Function]}
        placeholder="Search..."
        type="search"
        value=""
      />
    </header>
    <main
      className="cardList"
    >
      <UserCard
        isDisabled={false}
        onClick={[Function]}
        user={
          Object {
            "address": "minsk",
            "description": "Multi-layered client-server neural-net",
            "dob": "1990-01-01T00:00:00.000Z",
            "id": 91,
            "image": "https://images.unsplash.com/photo-1626038135427-bd4eb8193ba5?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
            "name": "Leanne Grahama",
          }
        }
      />
      <UserCard
        isDisabled={false}
        onClick={[Function]}
        user={
          Object {
            "address": "South Christy, Norberto Crossing",
            "description": "Synchronised bottom-line interface",
            "dob": "1990-01-01T00:00:00.000Z",
            "id": 92,
            "image": "https://images.unsplash.com/photo-1601455763557-db1bea8a9a5a?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
            "name": "Mrs. Dennis Schulist",
          }
        }
      />
    </main>
    <footer
      className="footer"
    >
      <Button
        id="loadMoreButton"
        isDisabled={false}
        onClick={[Function]}
      >
        LOAD MORE
      </Button>
    </footer>
  </div>
</Layout>
`;

exports[`should render overlay 1`] = `
<Layout
  className="layout"
  isAbsolute={true}
>
  <div
    className="base base__withOverlay"
  >
    <header
      className="header"
    >
      <p
        className="title"
      >
        Users list
      </p>
      <Input
        className="searchInput"
        onChange={[Function]}
        placeholder="Search..."
        type="search"
        value=""
      />
    </header>
    <main
      className="cardList"
    >
      <UserCard
        isDisabled={false}
        onClick={[Function]}
        user={
          Object {
            "address": "minsk",
            "description": "Multi-layered client-server neural-net",
            "dob": "1990-01-01T00:00:00.000Z",
            "id": 91,
            "image": "https://images.unsplash.com/photo-1626038135427-bd4eb8193ba5?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
            "name": "Leanne Grahama",
          }
        }
      />
      <UserCard
        isDisabled={false}
        onClick={[Function]}
        user={
          Object {
            "address": "South Christy, Norberto Crossing",
            "description": "Synchronised bottom-line interface",
            "dob": "1990-01-01T00:00:00.000Z",
            "id": 92,
            "image": "https://images.unsplash.com/photo-1601455763557-db1bea8a9a5a?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
            "name": "Mrs. Dennis Schulist",
          }
        }
      />
    </main>
    <footer
      className="footer"
    >
      <Button
        id="loadMoreButton"
        isDisabled={false}
        onClick={[Function]}
      >
        LOAD MORE
      </Button>
    </footer>
    <EditUserInfoOverlay
      hasScroll={false}
      onClose={[Function]}
      user={
        Object {
          "address": "minsk",
          "description": "Multi-layered client-server neural-net",
          "dob": "1990-01-01T00:00:00.000Z",
          "id": 91,
          "image": "https://images.unsplash.com/photo-1626038135427-bd4eb8193ba5?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
          "name": "Leanne Grahama",
        }
      }
    />
  </div>
</Layout>
`;

exports[`should render overlay with scroll 1`] = `
<Layout
  className="layout"
  isAbsolute={true}
>
  <div
    className="base base__withOverlay"
  >
    <header
      className="header"
    >
      <p
        className="title"
      >
        Users list
      </p>
      <Input
        className="searchInput"
        onChange={[Function]}
        placeholder="Search..."
        type="search"
        value=""
      />
    </header>
    <main
      className="cardList"
    >
      <UserCard
        isDisabled={false}
        onClick={[Function]}
        user={
          Object {
            "address": "minsk",
            "description": "Multi-layered client-server neural-net",
            "dob": "1990-01-01T00:00:00.000Z",
            "id": 91,
            "image": "https://images.unsplash.com/photo-1626038135427-bd4eb8193ba5?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
            "name": "Leanne Grahama",
          }
        }
      />
      <UserCard
        isDisabled={false}
        onClick={[Function]}
        user={
          Object {
            "address": "South Christy, Norberto Crossing",
            "description": "Synchronised bottom-line interface",
            "dob": "1990-01-01T00:00:00.000Z",
            "id": 92,
            "image": "https://images.unsplash.com/photo-1601455763557-db1bea8a9a5a?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
            "name": "Mrs. Dennis Schulist",
          }
        }
      />
    </main>
    <footer
      className="footer"
    >
      <Button
        id="loadMoreButton"
        isDisabled={false}
        onClick={[Function]}
      >
        LOAD MORE
      </Button>
    </footer>
    <EditUserInfoOverlay
      hasScroll={true}
      onClose={[Function]}
      user={
        Object {
          "address": "minsk",
          "description": "Multi-layered client-server neural-net",
          "dob": "1990-01-01T00:00:00.000Z",
          "id": 91,
          "image": "https://images.unsplash.com/photo-1626038135427-bd4eb8193ba5?ixid=MnwyODEwMzB8MHwxfHJhbmRvbXx8fHx8fHx8fDE2MzkwNTc3NTc&ixlib=rb-1.2.1&fm=jpg&w=168&h=168&fit=crop",
          "name": "Leanne Grahama",
        }
      }
    />
  </div>
</Layout>
`;
